manage/config-dialog.js

```diff
@@ -285,8 +312,50 @@ function configDialog(style) {
     }
   }
 
+  function updateVarOnBlur(e = {}) {
+    const value = Number(this.value);
+    const clamped = clampValue(value, this.va);
+    if (clamped === value) {
+      this.va.value = value;
+    }
+    if (e.type === 'blur') {
+      this.value = clamped;
```

Comment:
* You don't have to re-clamp the value when blurred, because the correct value is already set in the change handler.
* If the input value is invalid, `this.value` will not match `this.va.value` after blurred.
* In fact, we don't even need `clampValue` if we don't want to convert an invalid value to a correct one.

```diff
diff --git a/manage/config-dialog.js b/manage/config-dialog.js
index c314f2e..69cd88a 100644
--- a/manage/config-dialog.js
+++ b/manage/config-dialog.js
@@ -266,7 +266,8 @@ function configDialog(style) {
             onfocus: va.type === 'number' ? selectAllOnFocus : null,
             onblur: va.type === 'number' ? updateVarOnBlur : null,
             onchange: updateVarOnChange,
-            oninput: updateVarOnInput
+            oninput: updateVarOnInput,
+            required: true
           };
           if (typeof va.min === 'number') {
             options.min = va.min;
@@ -312,15 +313,8 @@ function configDialog(style) {
     }
   }
 
-  function updateVarOnBlur(e = {}) {
-    const value = Number(this.value);
-    const clamped = clampValue(value, this.va);
-    if (clamped === value) {
-      this.va.value = value;
-    }
-    if (e.type === 'blur') {
-      this.value = clamped;
-    }
+  function updateVarOnBlur() {
+    this.value = this.va.value;
   }
 
   function updateVarOnChange() {
@@ -328,29 +322,14 @@ function configDialog(style) {
       this.va.value = Number(this.value);
       updateRangeCurrentValue(this.va, this.va.value);
     } else if (this.type === 'number') {
-      updateVarOnBlur.call(this);
+      if (this.reportValidity()) {
+        this.va.value = Number(this.value);
+      }
     } else {
       this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
     }
   }
 
-  // Clamp input[type=number] to a valid range
-  function clampValue(value, va) {
-    // Don't restrict to integer values if step is undefined.
-    if (typeof va.step === 'number') {
-      const step = va.step || 1;
-      const scale = 10 ** (step.toString().split('.')[1] || '').length;
-      value = Math.round((value * scale) / (step * scale)) * (step * scale) / scale;
-    }
-    if (typeof va.min === 'number' && value < va.min) {
-      return va.min;
-    }
-    if (typeof va.max === 'number' && value > va.max) {
-      return va.max;
-    }
-    return value;
-  }
-
   function updateRangeCurrentValue(va, value) {
     const span = $('.current-value', va.input.closest('.config-range'));
     if (span) {
```

Reply:
`updateVarOnBlur` is wrong. The correct one:
```js
  function updateVarOnBlur() {
    this.value = isDefault(this.va) ? this.va.default : this.va.value;
  }
```

Reply:
I didn't know about `reportValidity()`... very cool. Of course it doesn't do anything in Firefox.

Reply:
Firefox:
![image](https://user-images.githubusercontent.com/1324510/45809867-e2a3d100-bcfb-11e8-8a3d-8ce482ea2a56.png)
Chrome:
![image](https://user-images.githubusercontent.com/1324510/45809926-01a26300-bcfc-11e8-878e-efde5aee3fb5.png)